### PR TITLE
 ⚠️ fix the leader election precedence over config file

### PR DESCRIPTION
**Pull request by aayushrangwala**
Fixes: #1463 

According to the documentation, flags to the cli should be considered over the properties in config files. But in the case of `--enable-leader-election` flag, there was a corner case which was not compliant to the statement.

case: If the user explicitly gives `false` to the leader election flag, but config file has true, the manager option is actually having the config value. Because of the fact that since it is a bool variable, its difficult to differentiate if the flag is given with value false or the zero value of the variable is being considered

Update LeaderElection Manager Option type to *bool

- Changed the datatype of leader election for manager option from bool to *bool

- Changed the flag parsing logic in Webhook main according to the new datatype
If the flag is not passed, it will be nil, otherwise false or true

- Added test to check the condition if true in config file and false in flag

### CODE REVIEW

**Comment by alvaroaleman** [comment]
Looks good, but is a breaking change.
/retitle :bug: fix the leader election precedence over config file
/ok-to-test

@vincepri WDYT, should we merge this or wait for a 0.9 release?

**Comment by alvaroaleman** [comment]
wow, fail.
/retitle :warning: fix the leader election precedence over config file

**Comment by vincepri** [comment]
Seems the code is now panicking.

**Comment by christopherhein** [comment]
These changes look good to me, thanks for catching that!

**Comment by aayushrangwala** [comment]
> Seems the tests are failing because of a panic, probably due to the pointer change.

@vincepri Yes, I handled the invalid pointer error. Tests seems to be passing now

File: pkg/webhook/conversion/testdata/main.go
Comment (christopherhein) [inline pkg/webhook/conversion/testdata/main.go:61]: I wonder if we could provide a helper for this since this is a big chunk people will have to repeat.

Comment (aayushrangwala) [reply pkg/webhook/conversion/testdata/main.go:61]: I think the main code lines the users will repeat is to convert normal bool var to pointer which can be done by k8s util `pointer.BoolPtr()`. If we think about it in a generic way this function is taking the string and assigning/returning the bool ptr and error `IF` its a valid string. If user wants that they can directly `strong.ParseBool()`

Or I am I missing some point?

Comment (vincepri) [reply pkg/webhook/conversion/testdata/main.go:61]: Why do we need to change the flag?

Comment (vincepri) [reply pkg/webhook/conversion/testdata/main.go:61]: We can probably just use the plain bool, and dereference it, or use pointer.BoolPtr(enableLeaderElection) instead.

Comment (aayushrangwala) [reply pkg/webhook/conversion/testdata/main.go:61]: That was the whole point for the change actually. If we keep the flag as plain bool, we won't be able to differentiate if the user explicitly gave it false or it was the zero value of that variable.

If user gave it false then we have to pass it down to the `Options` so that the config file option is not considered. If user did not give anything for this flag, then we will consider the config file option

Comment (christopherhein) [reply pkg/webhook/conversion/testdata/main.go:61]: I was more thinking of a helper function so every controller that currently uses:

```golang
flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
```

is now going to have to repeat your flap parsing func, while not hard it's breaking and something folks will have to do when upgrading, thus it could be nice to have a helper func people could define here instead. Something like:

```golang
var enableLeaderElection *bool
flag.Func("enable-leader-election",
		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.",
		ctrl.BoolPointerFlag(enableLeaderElection),
)
```

then in `pkg/manager` we could have `BoolPointerFlag()` as a helper.

```golang
func BoolPointerFlag(value *bool) func(flagVal string) error {
	return func(flagVal string) error {
		boolVal, err := strconv.ParseBool(flagVal)
		if err != nil {
			return err
		}
		value = &boolVal
		return nil
	}
}
```

Might be unnecessary 🤷‍♂️

Comment (vincepri) [reply pkg/webhook/conversion/testdata/main.go:61]: Does returning an error from BoolPointerFlag actually do what you want this to do? See https://play.golang.org/p/-VBOv0ZkZ1k, if the string is empty, the bool conversion fails

Comment (aayushrangwala) [reply pkg/webhook/conversion/testdata/main.go:61]: @vincepri Yeah I understood your point. So the current behaviour would be following:
- If `--enable-leader-election` is not passed -> the function block to pointer conversion won't run
- If `--enable-leader-election` is passed but with either empty string `""` or with some invalid string -> It will fail with showing the usage message
- If `--enable-leader-election` is passed with either `true/false` -> it will accept and move ahead

Comment (aayushrangwala) [reply pkg/webhook/conversion/testdata/main.go:61]: @christopherhein I got your point. I will add that helper

File: pkg/client/apiutil/apimachinery.go
Comment (vincepri) [inline pkg/client/apiutil/apimachinery.go:181]: Shouldn't we say `if flagVal == "" { return nil }` ?

Comment (aayushrangwala) [reply pkg/client/apiutil/apimachinery.go:181]: I think instead of returning nil, we should assign nil to the bool pointer. Which would imply that
- If the user gave the flag but with empty value, it will be considered as flag not passed
- If the user gave the flag but with some invalid value, it will be error
-  If the user gave the flag but with some valid value, it will be handled properly

**Comment by aayushrangwala** [comment]
@vincepri @christopherhein  Please approve if LG

**Comment by christopherhein** [comment]
This LGTM.

**Comment by aayushrangwala** [comment]
@vincepri Can you please take a final look and add approved/lgtm label. Since this is api breaking change we can have it in the next patch release

**Review by alvaroaleman — APPROVED** [review]
/hold

File: pkg/client/apiutil/apimachinery.go
Comment (alvaroaleman) [inline pkg/client/apiutil/apimachinery.go:191]: Actually, this doesn't do what you think it does. The `ptr` variable holds a pointer and is scoped to this function. You will need to:
* Set its value (which is problematic here I think, because then it will always be non-nil?)
* Use a double pointer `**bool`
/lgtm cancel

File: pkg/manager/manager_test.go
Comment (alvaroaleman) [inline pkg/manager/manager_test.go:540]: Somewhat nitpicky, but if we import the `utils/pointer` image anyways, why not use `pointer.Boolptr` everywhere? It makes the code IMHO easier to read as you don't have to go back to check what the value of `leaderElectOpt` is

Comment (aayushrangwala) [reply pkg/manager/manager_test.go:540]: Agree. Changed

File: pkg/client/apiutil/apimachinery.go
Comment (aayushrangwala) [reply pkg/client/apiutil/apimachinery.go:191]: Okay, got it. Let me think and work on it. Thanks for pointing that out

**Comment by alvaroaleman** [comment]
@aayushrangwala will you be able to rebase and finish this?

**Pull request closed without merging**
